hail/src/main/scala/is/hail/types/physical/stypes/concrete/SStructOfArrays.scala

```diff
@@ -0,0 +1,144 @@
+import is.hail.asm4s.{Code, IntInfo, LongInfo, Settable, TypeInfo, Value, coerce}
+import is.hail.expr.ir.{EmitCode, EmitCodeBuilder, IEmitCode}
+import is.hail.types.physical.PType
+import is.hail.types.physical.stypes.{EmitType, SCode, SSettable, SType}
+import is.hail.types.physical.stypes.interfaces.{SContainer, SIndexableCode, SIndexableValue}
+import is.hail.types.virtual.{TArray, TBaseStruct, Type}
+import is.hail.utils.FastIndexedSeq
+
+case class SStructOfArrays(virtualType: TArray, elementsRequired: Boolean, fields: IndexedSeq[SContainer]) extends SContainer {
+  require(virtualType.elementType.isInstanceOf[TBaseStruct])
+  require(fields.forall(st => st.virtualType.isInstanceOf[TArray]))
+
+  private def structVirtualType: TBaseStruct = virtualType.elementType.asInstanceOf
+
+  override val elementType: SStackStruct = SStackStruct(structVirtualType, fields.map(_.elementEmitType))
+
+  def elementEmitType: EmitType = EmitType(elementType, elementsRequired)
+
+  protected[stypes] def _coerceOrCopy(cb: EmitCodeBuilder, region: Value[Region], value: SCode, deepCopy: Boolean): SCode = ???
+
+  private lazy val codeStarts = fields.map(_.nCodes).scanLeft(baseTupleTypes.length)(_ + _).init
+  private lazy val settableStarts = fields.map(_.nSettables).scanLeft(baseTupleTypes.length)(_ + _).init
+  private lazy val baseTupleTypes: IndexedSeq[TypeInfo[_]] =
+    IndexedSeq(IntInfo) ++ (if (elementsRequired) None else Some(LongInfo))
+
+  lazy val codeTupleTypes: IndexedSeq[TypeInfo[_]] = baseTupleTypes ++ fields.flatMap(_.codeTupleTypes())
+
+  override lazy val settableTupleTypes: IndexedSeq[TypeInfo[_]] = baseTupleTypes ++ fields.flatMap(_.settableTupleTypes())
+
+  def fromCodes(codes: IndexedSeq[Code[_]]): SCode = {
+    new SStructOfArraysCode(this,
+      coerce(codes(0)),
+      if (elementsRequired) None else Some(coerce(codes(1))),
+      fields.zipWithIndex.map { case (f, i) =>
+        val start = codeStarts(i)
+        f.fromCodes(codes.slice(start, start + f.nCodes)).asIndexable
+      }
+    )
+  }
+
+  def fromSettables(settables: IndexedSeq[Settable[_]]): SSettable = {
+    new SStructOfArraysSettable(this,
+      coerce(settables(0)),
+      if (elementsRequired) None else Some(coerce(settables(1))),
+      fields.zipWithIndex.map { case (f, i) =>
+        val start = settableStarts(i)
+        f.fromCodes(settables.slice(start, start + f.nSettables)).asInstanceOf
+      }
+    )
+  }
+
+
+  def canonicalPType(): PType = PType.canonical(virtualType)
+
+  def castRename(t: Type): SType = {
+    val arrayType = t.asInstanceOf[TArray]
+    val structType = arrayType.elementType.asInstanceOf[TBaseStruct]
+
+    SStructOfArrays(
+      arrayType,
+      elementsRequired,
+      structType.types.zip(fields).map { case (v, f) => f.castRename(TArray(v)).asInstanceOf }
+    )
+  }
+}
+
+class SStructOfArraysSettable(
+  val st: SStructOfArrays,
+  val length: Settable[Int],
+  val missing: Option[Settable[Long]],
```

Comment:
If the elements are optional and the field is optional, I intend to have the field missingness be an `OR` of these missing bits and the array's missing bits. That way accessing a single field can be entirely local to that field.